pkg/loop/internal/chain_reader.go

```diff
@@ -0,0 +1,134 @@
+const ParamsCurrentEncodingVersion = CBOREncodingVersion
+const RetvalCurrentEncodingVersion = CBOREncodingVersion
+
+func encodeVersionedBytes(data any, version int32) (*pb.VersionedBytes, error) {
+	var bytes []byte
+	var err error
+
+	switch version {
+	case JSONEncodingVersion1:
+		bytes, err = jsonv1.Marshal(data)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %w", types.ErrInvalidType, err)
+		}
+	case JSONEncodingVersion2:
+		bytes, err = jsonv2.Marshal(data)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %w", types.ErrInvalidType, err)
+		}
+	case CBOREncodingVersion:
+		enco := cbor.CoreDetEncOptions()
+		enco.Time = cbor.TimeRFC3339Nano
+		var enc cbor.EncMode
+		enc, err = enco.EncMode()
+		if err != nil {
+			return nil, err
+		}
+		bytes, err = enc.Marshal(data)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %w", types.ErrInvalidType, err)
+		}
+	default:
+		return nil, fmt.Errorf("unsupported encoding version %d for data %v", version, data)
+	}
+
+	return &pb.VersionedBytes{Version: uint32(version), Data: bytes}, nil
+}
+
+func decodeVersionedBytes(res any, vData *pb.VersionedBytes) error {
+	var err error
+	switch vData.Version {
+	case JSONEncodingVersion1:
+		err = jsonv1.Unmarshal(vData.Data, res)
+	case JSONEncodingVersion2:
+		err = jsonv2.Unmarshal(vData.Data, res)
+	case CBOREncodingVersion:
+		err = cbor.Unmarshal(vData.Data, res)
+	default:
+		return fmt.Errorf("unsupported encoding version %d for versionedData %v", vData.Version, vData.Data)
+	}
+
+	if err != nil {
+		return fmt.Errorf("%w: %w", types.ErrInvalidType, err)
+	}
+	return nil
+}
+
+func (c *chainReaderClient) GetLatestValue(ctx context.Context, bc types.BoundContract, method string, params, retVal any) error {
+	versionedParams, err := encodeVersionedBytes(params, ParamsCurrentEncodingVersion)
+	if err != nil {
+		return err
+	}
+
+	boundContract := pb.BoundContract{Name: bc.Name, Address: bc.Address, Pending: bc.Pending}
+
+	reply, err := c.grpc.GetLatestValue(ctx, &pb.GetLatestValueRequest{Bc: &boundContract, Method: method, Params: versionedParams})
+	if err != nil {
+		return types.UnwrapClientError(err)
+	}
+
+	return decodeVersionedBytes(retVal, reply.RetVal)
+}
+
+var _ pb.ChainReaderServer = (*chainReaderServer)(nil)
+
+type chainReaderServer struct {
+	pb.UnimplementedChainReaderServer
+	impl types.ChainReader
+}
+
+func (c *chainReaderServer) GetLatestValue(ctx context.Context, request *pb.GetLatestValueRequest) (*pb.GetLatestValueReply, error) {
+	var bc types.BoundContract
+	bc.Name = request.Bc.Name[:]
+	bc.Address = request.Bc.Address[:]
+	bc.Pending = request.Bc.Pending
+
+	params := &map[string]any{}
+
+	if err := decodeVersionedBytes(params, request.Params); err != nil {
+		return nil, err
+	}
+
+	retVal := &map[string]any{}
+	err := c.impl.GetLatestValue(ctx, bc, request.Method, params, retVal)
+	if err != nil {
+		return nil, err
+	}
+
+	encodedRetVal, err := encodeVersionedBytes(retVal, RetvalCurrentEncodingVersion)
```

Comment:
Shouldn't we use the same version that they used to encode the params?  This would ensure that the client supports the version.  They should always be the same, but just in case.